PurchasesCoreSwift/Identity/PurchaserInfoManager.swift

```diff
@@ -0,0 +1,191 @@
+
+    @objc public weak var delegate: PurchaserInfoManagerDelegate?
+
+    private(set) var lastSentPurchaserInfo: PurchaserInfo?
+    private let operationDispatcher: OperationDispatcher
+    private let deviceCache: DeviceCache
+    private let backend: Backend
+    private let systemInfo: SystemInfo
+    private let purchaserInfoCacheLock = NSRecursiveLock()
+
+    @objc public init(operationDispatcher: OperationDispatcher,
+                      deviceCache: DeviceCache,
+                      backend: Backend,
+                      systemInfo: SystemInfo) {
+        self.operationDispatcher = operationDispatcher
+        self.deviceCache = deviceCache
+        self.backend = backend
+        self.systemInfo = systemInfo
+    }
+
+    @objc public func fetchAndCachePurchaserInfo(appUserID: String,
+                                                 isAppBackgrounded: Bool,
+                                                 completion: ReceivePurchaserInfoBlock?) {
+        deviceCache.setPurchaserInfoCacheTimestampToNow(appUserID: appUserID)
+        operationDispatcher.dispatchOnWorkerThread(withRandomDelay: isAppBackgrounded) {
+            self.backend.getSubscriberData(appUserID: appUserID) { purchaserInfo, error in
+                if error != nil {
+                    self.deviceCache.clearPurchaserInfoCacheTimestamp(appUserID: appUserID)
+                } else if let info = purchaserInfo {
+                    self.cache(purchaserInfo: info, appUserID: appUserID)
+                }
+
+                if let completion = completion {
+                    self.operationDispatcher.dispatchOnMainThread {
+                        completion(purchaserInfo, error)
+                    }
+                }
+
+            }
+        }
+    }
+
+    @objc public func fetchAndCachePurchaserInfoIfStale(appUserID: String,
+                                                        isAppBackgrounded: Bool,
+                                                        completion: ReceivePurchaserInfoBlock?) {
+        let cachedPurchaserInfo = cachedPurchaserInfo(appUserID: appUserID)
+        let isCacheStale = deviceCache.isPurchaserInfoCacheStale(appUserID: appUserID,
+                                                                 isAppBackgrounded: isAppBackgrounded)
+        let needsToRefresh = isCacheStale || cachedPurchaserInfo == nil
+        if needsToRefresh {
+            Logger.debug(isAppBackgrounded
+                            ? Strings.purchaserInfo.purchaserinfo_stale_updating_in_background
+                            : Strings.purchaserInfo.purchaserinfo_stale_updating_in_foreground)
+            fetchAndCachePurchaserInfo(appUserID: appUserID,
+                                       isAppBackgrounded: isAppBackgrounded,
+                                       completion: completion)
+            Logger.rcSuccess(Strings.purchaserInfo.purchaserinfo_updated_from_network)
```

Comment:
this log doesn't _sound_ like it'd be specific to this method, but rather to all fetching? and as for our potential new "rule", should this log be as close the operation as possible, i.e. in the `fetchAndCachePurchaserInfo` method?

Reply:
Agreed, updating.